jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -50,7 +50,31 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        write(Content.Chunk.from(byteBuffer, last, callback::succeeded), callback);
+        // This class' read() impl calls succeeded() on the callback before returning it to the
+        // caller, so when EOF / EMPTY are used this is perfectly right as the chunk can be
+        // considered consumed and there is no buffer to reclaim. But when a proper buffer is
+        // used the passed callback must be wrapped to make its succeeded() method a noop
+        // as it is the release of the chunk that should trigger the success of the callback.
+        Content.Chunk chunk;
+        Callback cb;
+        if (!byteBuffer.hasRemaining())
+        {
+            chunk = last ? Content.Chunk.EOF : Content.Chunk.EMPTY;
+            cb = callback;
+        }
+        else
+        {
+            chunk = Content.Chunk.from(byteBuffer, last, callback::succeeded);
+            cb = new Callback.Nested(callback)
+            {
+                @Override
+                public void succeeded()
+                {
+                    // Noop, chunk.release() will trigger the callback's success.
+                }
+            };
+        }
+        write(chunk, cb);
```

Comment:
I would do the other way around.
Have an `private offer()` method called by the 2 `write()` methods, so that you can release the chunk when the read releases the chunk.